exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -307,3 +315,235 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
+		return errors.Wrapf(err, "failed to create machines for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	}
+
+	return nil
+}
+
+// deleteMachinesWithoutInfrastructure deletes all MachinePool Machines with an infraRef pointing to an infraMachine that no longer exists.
+// This allows us to clean up Machines when the MachinePool is scaled down.
+func (r *MachinePoolReconciler) deleteMachinesWithoutInfrastructure(ctx context.Context, mp *expv1.MachinePool, machines []clusterv1.Machine, infraMachines []unstructured.Unstructured) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	log.V(2).Info("Deleting machines without infrastructure", "machinePool", mp.Name, "namespace", mp.Namespace)
+
+	infraMachineNames := sets.Set[string]{}
+	for _, infraMachine := range infraMachines {
+		infraMachineNames.Insert(infraMachine.GetName())
+	}
+
+	for i := range machines {
+		machine := &machines[i]
+		infraRef := machine.Spec.InfrastructureRef
+
+		// Check this here since external.Get() will fail if the infraRef is empty, i.e. no object kind or apiVersion.
+		// Delete the machine if the infraRef is empty.
+		if (infraRef == corev1.ObjectReference{}) {
+			log.V(4).Info("Machine has an empty infraRef, will delete", "machine", machine.Name, "namespace", machine.Namespace)
+			if err := r.Client.Delete(ctx, machine); err != nil {
+				return err
+			}
+			continue
+		}
+
+		if _, ok := infraMachineNames[infraRef.Name]; !ok {
+			log.V(4).Info("Deleting machine without infrastructure", "machine", machine.Name, "namespace", machine.Namespace)
+			if err := r.Client.Delete(ctx, machine); err != nil {
+				return err
+			}
+		} else {
+			log.V(4).Info("Machine has infrastructure, nothing to do", "machine", machine.Name, "namespace", machine.Namespace)
+		}
+	}
+
+	return nil
+}
+
+// createMachinesIfNotExists creates a MachinePool Machine for each infraMachine if it doesn't already exist and sets the owner reference and infraRef.
+func (r *MachinePoolReconciler) createMachinesIfNotExists(ctx context.Context, mp *expv1.MachinePool, machines []clusterv1.Machine, infraMachines []unstructured.Unstructured) ([]clusterv1.Machine, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Construct a set of names of infraMachines that already have a Machine.
+	infraRefNames := sets.Set[string]{}
+	for _, machine := range machines {
+		infraRef := machine.Spec.InfrastructureRef
+		infraRefNames.Insert(infraRef.Name)
+	}
+
+	for i := range infraMachines {
+		infraMachine := &infraMachines[i]
+		// If infraMachine already has a Machine, skip it.
+		if infraRefNames.Has(infraMachine.GetName()) {
+			continue
+		}
+
+		// Otherwise create a new Machine for the infraMachine.
+		log.V(4).Info("Creating new Machine for infraMachine", "infraMachine", infraMachine.GetName(), "namespace", infraMachine.GetNamespace())
+		machine := getNewMachine(mp, infraMachine)
+		if err := r.Client.Create(ctx, machine); err != nil {
```

Comment:
```suggestion
		if err := r.Client.Create(ctx, machine); err != nil && !apierros.IsAleadyExists(err) {
```
To make it resilient to stale cache reads
For the same reason, it would be great to wait for the object to exists in cache (call get until it returns the object)

Reply:
I think this won't help if we create a Machine with generate name

Reply:
I think the isalreadyexists won't help with generate name. Waiting for the cache should help though

Reply:
Do we still want to add the API error check then or something else?

Reply:
I think we only would want something like this: https://github.com/kubernetes-sigs/cluster-api/blob/f37c32984f741281cb8aba0f6ff5a3f9a1bd2ef0/internal/controllers/machineset/machineset_controller.go#L724-L739

Reply:
I'm a bit confused with what the MachineSet controller is trying to do. It looks like we're trying to create the Machines first and then adding this wait function in the return statement, and I'm also not sure what `ssa.Patch` is compared to `client.Create()`. Also, would this work with `GenerateName` still or would we need to specify the machine.Name field first?

Reply:
It should work with GenerateName. The machine object should contain the Name set after the Create call. (controller-runtime updates the object based on the response from the apiserver).

Basically the idea is:
* create the machine
* wait until client.Get sees the Machine
  * client.Get / List are both using a local cache

Waiting until we can get the Machine in the cache ensures that any subsequent reconcile will be able to see the Machine. Otherwise it can happen that you create the Machine, but the next reconcile won't see it  (as result of the List call) because the local cache is stale

Reply:
I think I got it, thanks! 

Note: for anyone w/o context, we talked a bit more about this in Slack